### PR TITLE
feat(taps): add EndOfStreamError to skip partitions and streams gracefully

### LINKED ISSUE #280: Let taps continue to next partitions and/or streams upon certain errors
Migrated from GitLab: https://gitlab.com/meltano/sdk/-/issues/282

Originally created by @LaurentS on 2021-11-29 22:40:06

---

## Summary
[//]: # (Concisely summarize the feature you are proposing.)
Follow up to [!195](https://gitlab.com/meltano/sdk/-/merge_requests/195) and probably related to [#137](https://gitlab.com/meltano/sdk/-/issues/137) (and maybe [#134](https://gitlab.com/meltano/sdk/-/issues/134))

In release 0.3.14, there is now support for `RetriableAPIError` and `FatalAPIError`, which I think should be extended to allow for the tap to continue after such an error (whether that's the default or an opt-in behaviour), in particular when dealing with a partitioned stream.

## Proposed benefits
[//]: # (Concisely summarize the benefits this feature would bring to yourself and other users.)
Example use case: `tap-github` can fetch data for a list of repos. Currently, if it fails at some point, the rest of the list is not handled at all. See [this issue](https://github.com/MeltanoLabs/tap-github/issues/52) for more details. In that particular case, not handling the error means we're not able to retrieve data past the error which seems persistent on some repos.

Given that the tap is run as a separate process, bubbling the exception up to the top level doesn't actually help in that the calling process cannot get access to the details of it. Returning state (and maybe an error exit code) seems like a cleaner way to fail.

## Proposal details
[//]: # (In as much detail as you are able, describe the feature you'd like to build or would like to see built.)

In `Stream._sync_records`, the code [loops through a list of contexts](https://gitlab.com/meltano/sdk/-/blob/main/singer_sdk/streams/core.py#L911), but there is no error handling. The loop could be enclosed in a `try...except` block which could selectively capture the above exceptions and update state (or not) accordingly, before proceeding with the next context.

## Best reasons not to build
[//]: # (Will this negatively affect any existing functionality? Do you anticipate any breaking changes versus what may already be working today? Make the counter-argument to your proposal here.)

I can't really think of a reason why this would cause problems, but I only have the `tap-github` use case in mind. If there are downsides, that behaviour could be opt-in via a config option, with the default behaviour remaining as is.

### PR DESCRIPTION
**Pull request by daria-hrabar**
Closes #280

## Summary

Introduces `EndOfStreamError`, a new exception in `singer_sdk/exceptions.py` that tap developers can raise inside `get_records()` to signal that a partition or stream should be skipped without aborting the entire sync. The SDK catches this signal — along with all other `SkippableSyncError` subclasses — at two levels: inside the `get_records` iteration in `Stream._sync_records` and around each stream call in `Tap.sync_all`. On a skippable error, a warning is logged, partition state is finalized, and the sync continues to the next partition or stream.

## Motivation

Currently, any unhandled exception raised during a partitioned sync aborts the entire tap run. All remaining partitions and streams are silently skipped, and no state is emitted for them. This is particularly painful for taps like `tap-github` that iterate over a list of repositories as partitions — a single persistent API error on one repository prevents all subsequent repositories from being synced (see [MeltanoLabs/tap-github#52](https://github.com/MeltanoLabs/tap-github/issues/52)).

## Changes

- `singer_sdk/exceptions.py` — added `EndOfStreamError(SkippableSyncError)` and registered it in `__all__`
- `singer_sdk/streams/core.py` — wrapped `get_records()` and its iteration loop in `Stream._sync_records` with `try/except SkippableSyncError` (skip, finalize state, continue) and `except RetriableSyncError` (finalize state, re-raise); added `RetriableSyncError` and `SkippableSyncError` to imports; updated docstring
- `singer_sdk/tap_base.py` — wrapped `stream.sync()` in `Tap.sync_all` with `try/except SkippableSyncError`, logging a warning and calling `continue`; added `SkippableSyncError` to imports
- `tests/core/test_streams.py` — added 11 unit tests covering partition-level skip, stream-level skip, exception hierarchy, edge cases (empty partition list, empty stream list, all/first/last/only partition failing), and verification that non-`SkippableSyncError` exceptions still propagate uncaught at both levels

## Design decisions

- `EndOfStreamError` is an **explicit, opt-in signal** for tap developers. Raising it inside `get_records()` signals that a partition is intentionally skippable. Plain exceptions (e.g. `RuntimeError`) still crash the tap — this is intentional, as they may indicate genuine bugs rather than expected, recoverable conditions.
- The catch uses `SkippableSyncError` rather than `EndOfStreamError` directly, so existing subclasses such as `SkippableAPIError` and `InvalidRecord` also benefit from the skip-and-continue behavior without requiring tap code changes.
- `RetriableSyncError` is caught at the partition level to finalize and flush state before re-raising, ensuring the orchestrator has a clean checkpoint before the retry mechanism takes over. It is not swallowed.
- `FatalSyncError` and its subclasses (including `FatalAPIError`) are intentionally left uncaught. These signal conditions severe enough to abort the entire sync — such as missing key properties or invalid stream sort order — and silently skipping them would hide genuine data integrity issues.
- The `try` block wraps both the `get_records()` call and the iteration loop over it. This is necessary because `get_records` is a generator function — calling it does not execute any of its code. The `EndOfStreamError` is only raised during iteration, not at the point of the call.
- The stream-level catch in `Tap.sync_all` is separate from the partition-level catch in `_sync_records` because a stream can raise `SkippableSyncError` before partition iteration begins — both levels are needed to cover different failure points in the sync lifecycle.
- A structured error-reporting mechanism for skipped partitions (beyond warning logs) is not included in this PR. Input from maintainers on the preferred shape of such a mechanism is welcome before implementation.

## Testing

All 855 tests pass across Python 3.10–3.14 with no regressions. `singer_sdk/exceptions.py` reaches 100% coverage.

## Summary by Sourcery

Allow taps to skip explicitly skippable partitions and streams while continuing synchronization and preserving state.

New Features:
- Add EndOfStreamError as an explicit signal for safely skipping unrecoverable partitions or streams during a sync.

Bug Fixes:
- Prevent skippable partition or stream failures from aborting the remaining synchronization work while preserving state checkpoints.
- Finalize and flush partition state before propagating retriable synchronization errors.

Enhancements:
- Handle all SkippableSyncError subclasses consistently at both partition and stream synchronization levels while allowing unrelated and fatal errors to propagate.

Tests:
- Add comprehensive coverage for partition and stream skipping, state-safe continuation, edge cases, exception inheritance, and propagation of non-skippable errors.

### CODE REVIEW

**Comment by daria-hrabar** [comment]
@sourcery-ai review

File: singer_sdk/tap_base.py [hunk 1]
```diff
@@ -510,8 +511,16 @@ def sync_all(self) -> None:
...
+            try:
+                stream.sync()
+                stream.finalize_state_progress_markers()
+            except SkippableSyncError as e:
+                self.logger.warning(
+                    "Skipping stream '%s': %s",
+                    stream.name,
+                    str(e),  # ruff: ignore[logging-eager-conversion]
+                )
+                continue
```
[the remark this answers is not in the document]
Comment (edgarrmondragon) [reply singer_sdk/tap_base.py:523]: @daria-hrabar Should we call `stream.finalize_state_progress_markers()` in a finally block?
